### PR TITLE
HTTP/2 procedure module

### PR DESCRIPTION
**Pull request by bmcelvee**
Related to https://github.com/openshift/openshift-docs/pull/23565

### CODE REVIEW

File: modules/nw-http2-haproxy.adoc [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+Enable HTTP/2 on a single Ingress Controller.
+
+* To enable HTTP/2 on an Ingress Controller, run the `oc annotate` command:
++
+----
+$ oc -n openshift-ingress-operator annotate ingresscontrollers/<ingresscontroller_name> ingress.operator.openshift.io/default-enable-http2=true
+----
+
+Enable HTTP/2 on all Ingress Controllers.
```
Comment (bmcelvee) [inline modules/nw-http2-haproxy.adoc:36]: Clarify that this is for the entire cluster

Comment (sferich888) [reply modules/nw-http2-haproxy.adoc:36]: +!

**Comment by bmcelvee** [comment]
@openshift/team-documentation please peer review, thanks!

File: modules/nw-http2-haproxy.adoc [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+The connection from HAProxy to the application Pod can use HTTP/2 only for
+re-encrypt routes and not for edge-terminated or insecure routes. This
+restriction comes from the fact that HAProxy uses Application-Level Protocol
+Negotiation (ALPN), which is a TLS extension, to negotiate the use of HTTP/2
+with the back-end. The implication is that end-to-end HTTP/2 is possible with
+passthrough and re-encrypt and not with insecure or edge-terminated routes.
+
+[IMPORTANT]
+====
+A connection that uses the HTTP/2 protocol cannot be upgraded to the WebSocket protocol. If you have a back-end application that is designed to allow WebSocket connections, it must not allow a connection to negotiate use of the HTTP/2 protocol or else WebSocket connections will fail.
```
Comment (Miciah) [inline modules/nw-http2-haproxy.adoc:27]: Iterating on this, we could use the following if we want to be more explicit and it's not too wordy:

> For non-passthrough routes, the IngressController negotiates its connection to the application independently of the connection from the client.  This means a client may connect to the IngressController and negotiate HTTP/1.1, and the IngressController may then connect to the application, negotiate HTTP/2, and forward the request from the client HTTP/1.1 connection using the HTTP/2 connection to the application.  This poses a problem if the client subsequently tries to upgrade its connection from HTTP/1.1 to the WebSocket protocol, because the IngressController cannot forward WebSocket to HTTP/2 and cannot upgrade its HTTP/2 connection to WebSocket.  Consequently, if you have an application that is intended to accept WebSocket connections, it must not allow negotiating the HTTP/2 protocol or else clients will fail to upgrade to the WebSocket protocol.

File: modules/nw-http2-haproxy.adoc [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+====
+
+.Procedure
+
+Enable HTTP/2 on a single Ingress Controller.
+
+* To enable HTTP/2 on an Ingress Controller, run the `oc annotate` command:
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:34]: I've been using _enter_ instead of _run_.

> ISG: To instruct users to enter commands, use type, enter, or issue. Use one term consistently through- out your information. To refer to the area where you enter commands, in documentation for only Windows operating systems, use command prompt. In documentation for other operating sys- tems or for both Windows and other operating systems, use command line. Do not use run to refer to specifying a command on a command line or at a command prompt. However, you can use run to refer to the general use of a command. (p. 186)

File: modules/nw-http2-haproxy.adoc [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+passthrough and re-encrypt and not with insecure or edge-terminated routes.
+
+[IMPORTANT]
+====
+A connection that uses the HTTP/2 protocol cannot be upgraded to the WebSocket protocol. If you have a back-end application that is designed to allow WebSocket connections, it must not allow a connection to negotiate use of the HTTP/2 protocol or else WebSocket connections will fail.
+====
+
+.Procedure
+
+Enable HTTP/2 on a single Ingress Controller.
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:32]: I think this preamble is optional, and I haven't been including one generally.

File: modules/nw-http2-haproxy.adoc [hunk 5]
```diff
@@ -0,0 +1,46 @@
+// Module included in the following assemblies:
+//
+// * networking/ingress-operator.adoc
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:3]: While it's outside the scope, I've long wanted to see this assembly split up into user stories for each of these (HTTP/2, sharding, ect.). I haven't had time to delve deeper into it.

File: modules/nw-http2-haproxy.adoc [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+[id="nw-http2-haproxy{context}"]
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:5]: "nw-http2-haproxy_{context}"

File: modules/nw-http2-haproxy.adoc [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+= HTTP/2 Ingress connectivity
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:6]: Is this more like _Enabling HTTP/2 Ingress connectivity_ or similar?

File: modules/nw-http2-haproxy.adoc (hunk 3 above)
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:34]: Add,

Replace `<ingresscontroller_name>` with the name of the Ingress Controller to annotate.

File: modules/nw-http2-haproxy.adoc [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+* To enable HTTP/2 on an Ingress Controller, run the `oc annotate` command:
++
+----
+$ oc -n openshift-ingress-operator annotate ingresscontrollers/<ingresscontroller_name> ingress.operator.openshift.io/default-enable-http2=true
+----
+
+Enable HTTP/2 on the entire cluster.
+
+* To enable HTTP/2 for the entire cluster, run the `oc annotate` command:
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:42]: s/run/enter/

File: modules/nw-http2-haproxy.adoc [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+Enable HTTP/2 on a single Ingress Controller.
+
+* To enable HTTP/2 on an Ingress Controller, run the `oc annotate` command:
++
+----
+$ oc -n openshift-ingress-operator annotate ingresscontrollers/<ingresscontroller_name> ingress.operator.openshift.io/default-enable-http2=true
+----
+
+Enable HTTP/2 on the entire cluster.
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:40]: Are these two distinct single step procedures? Enabling for a specific controller and cluster-wide seems mutually exclusive.

File: modules/nw-http2-haproxy.adoc [hunk 10]
```diff
@@ -0,0 +1,46 @@
+// Module included in the following assemblies:
+//
+// * networking/ingress-operator.adoc
+
+[id="nw-http2-haproxy{context}"]
+= HTTP/2 Ingress connectivity
+
+You can enable transparent end-to-end HTTP/2 connectivity in HAProxy. It allows application owners to make use of HTTP/2 protocol capabilities, including single connection, header compression, binary streams, and more.
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:8]: Is this too much conceptual exposition prior to a procedure? Right now there's probably nowhere else for it to land.

File: modules/nw-http2-haproxy.adoc [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+To enable the use of HTTP/2 for the connection from the client to HAProxy, a
+route must specify a custom certificate. A route that uses the default
+certificate cannot use HTTP/2. This restriction is necessary to avoid problems
+from connection coalescing, where the client re-uses a connection for different
+routes that use the same certificate.
+
+The connection from HAProxy to the application Pod can use HTTP/2 only for
+re-encrypt routes and not for edge-terminated or insecure routes. This
+restriction comes from the fact that HAProxy uses Application-Level Protocol
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:20]: s/This restriction comes from the fact that/This restriction is because/

File: modules/nw-http2-haproxy.adoc [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+Negotiation (ALPN), which is a TLS extension, to negotiate the use of HTTP/2
```
Comment (jboxman) [inline modules/nw-http2-haproxy.adoc:21]: @Miciah, if I'm honest, I have to admit I don't why ALPN means only passthrough and re-encrypt are permissible, and insecure and edge-terminated are not. But I haven't configured an HTTP proxy since v1.1 days, so it might be obvious.

**Review by jboxman — COMMENTED** [review]
@bmcelvee, left some comments. Thanks!

File: modules/nw-http2-haproxy.adoc (hunk 12 above)
Comment (frobware) [reply modules/nw-http2-haproxy.adoc:21]: Some of the limitation comes from plumbing the correct "server" line in haproxy. 

In https://github.com/openshift/router/pull/104 we were trying to make gRPC transparently work. If you have a client(gRPC)->haproxy(edge terminated)->backend you need to explicitly change the haproxy behaviour away from ALPN. What would normally be `alpn h2,http/1.1` now must be `proto h2`. The following link [1] details the `proto h2` reasoning. 

We really don't have complete information on what the backend does (or can do) to decide whether the config can default to `alpn h2` on edge-terminated routes. Substituting for `proto h2` by automatically sniffing for `application/grpc` in a header works for the specific gRPC case, but breaks down if you are multiplexing on the backend for, say, gRPC **and** HTTP protocols.

The TL;DR is we didn't find a generic 100% accurate mapping for the haproxy config. The current/future thinking is for the service/route to explicitly state its needs and plumb the haproxy config based on that. Overall, the limitation of only supporting re-encrypt and passthrough may be a limitation of configuring haproxy within the context of openshift; I haven't looked at other proxy implementations to see if this could be easier, or more transparent there.

[1] - https://www.haproxy.com/blog/haproxy-2-0-and-beyond/#end-to-end-http-2

File: modules/nw-http2-haproxy.adoc (hunk 7 above)
Comment (frobware) [reply modules/nw-http2-haproxy.adoc:6]: I think so; if you add the annotation you're explicitly enabling HTTP/2.

File: modules/nw-http2-haproxy.adoc [hunk 13]
```diff
@@ -0,0 +1,46 @@
+// Module included in the following assemblies:
+//
+// * networking/ingress-operator.adoc
+
+[id="nw-http2-haproxy{context}"]
+= HTTP/2 Ingress connectivity
+
+You can enable transparent end-to-end HTTP/2 connectivity in HAProxy. It allows application owners to make use of HTTP/2 protocol capabilities, including single connection, header compression, binary streams, and more.
+
+You can enable this HTTP/2 connectivity for an individual Ingress Controller or for the entire cluster.
```
Comment (frobware) [inline modules/nw-http2-haproxy.adoc:10]: Drop "this".

File: modules/nw-http2-haproxy.adoc (hunk 9 above)
Comment (Miciah) [reply modules/nw-http2-haproxy.adoc:40]: Right, they are 2 separate procedures, each procedure being 1 step long.

File: modules/nw-http2-haproxy.adoc (hunk 12 above)
Comment (Miciah) [reply modules/nw-http2-haproxy.adoc:21]: The reason why the use of ALPN means only it only works for passthrough and reencrypt routes is in the parenthetical ("which is a TLS extension"), but it certainly could be stated more explicitly.

Generally, HTTP/2 is designed for use with TLS, and HAProxy uses ALPN during the TLS handshake to negotiate the use of HTTP/2.  When one uses a passthrough route, HAProxy passes the client connection through to the application, the client and application perform the TLS handshake, and so they can negotiate HTTP/2 using ALPN.  When one uses a reencrypt route, HAProxy opens a TLS connection to the application, HAProxy and the application perform the TLS handshake, and so they likewise can negotiate HTTP/2 using ALPN.  In the case of insecure and edge-terminated routes however, HAProxy opens a cleartext HTTP connection to the application, so there is no TLS and therefore no TLS handshake, no ALPN, and no way to negotiate HTTP/2.

File: modules/nw-http2-haproxy.adoc (hunk 5 above)
Comment (bmcelvee) [reply modules/nw-http2-haproxy.adoc:3]: Same! I had intended to do just that with some of the larger 4.5 additions, but I think we should do some networking docs bucket coordination. :) Something to look forward to...soon. :)

File: modules/nw-http2-haproxy.adoc (hunk 9 above)
Comment (bmcelvee) [reply modules/nw-http2-haproxy.adoc:40]: Yeah, that's why I included to preamble bit.

File: modules/nw-http2-haproxy.adoc (hunk 10 above)
Comment (bmcelvee) [reply modules/nw-http2-haproxy.adoc:8]: Probably, it doesn't need to stand alone so I think it's okay for now. In later iterations (when we reorg networking content) this could become a conceptual module.

**Comment by bmcelvee** [comment]
@lihongan would you mind reviewing this PR, please? Thanks!

**Comment by lihongan** [comment]
Looks great! Thank you @bmcelvee
